Propose an ADR on reference scripts (#438)
input-output-hk/hydra: docs/adr/2022-07-22_019-reference-scripts.md

---
slug: 19
title: | 
  19. Use of reference scripts
authors: []
tags: [Proposed]
---

## Status

Proposed

## Context

* In the desire to make Hydra transactions smaller and cheaper (at the time of writing any abort tx was too big), we want to use the **reference script** and **reference input** features of the upcoming `Babbage` ledger era. See the [babbage ledger spec](https://hydra.iohk.io/build/16861604/download/1/babbage-changes.pdf), [CIP-31](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0031) and [CIP-33](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0033) for details.

> * With these features we not need to (re-)include the same scripts in each transaction.

Reviewer: ```suggestion
* With these features we do not need to (re-)include scripts in each transaction.
```

* The CIPs do not specify how reference scripts are to be managed and we can see at least two options:
  1. Add them as outputs to the `init` transaction or prior that as part of each Hydra Head instance
  2. Post them out-of-band, separate to individual Head instances

> * Ownership of the outputs holding the scripts is to be considered. If these "reference outputs" are spent, they cannot be referred to anymore. This would mean all heads referring ot them become denialed of service (DoS).

Reviewer: ```suggestion
* Ownership of the outputs holding the scripts is to be considered. If these "reference outputs" are spent, they cannot be referred to anymore. This would mean all heads referring to them can be denied of service (DoS).
```

* Each head will need to refer to the correct version of the hydra scripts. That is, consistent with the script hashes known to the `hydra-node`.
  + This is also related to the problem of managing script versions & updates.
  + Right now, the `hydra-node` is compiled against `hydra-plutus` to access compiled script content and hashes.

* The general trade-off is: instead of paying ADA fees for scripts adding to the transaction size in _each_ transaction, ADA deposits will need to be put down to have scripts be part of the UTxO set in the ledger _once_.

## Decision

* Publish outputs holding Hydra scripts out-of-band (option 2), because
  + All scripts would not fit into the `init` transaction directly, we would need to post multiple.
  + Costs (deposits) would need to be payed for each head instance.

>   + Although, heads using option 1 would not be exposed to DoS 

Reviewer: The way this is presented is a bit weird because, both option (1) and option (2) are about publishing Hydra scripts on-chain; the only question is whether or not to mix that with the init transaction. 

And the conclusion being that no, we'd rather not for reason X & Y. 

>    + Although, heads using option 1 would not be exposed to DoS 

I disagree, it would be exposed to it all-the-same. The DoS problem stems from what locks the outputs where the script is posted to. If it's spendable, then it's deniable. So these are kind of orthogonal. 

One extra reason why doing it as part of the init isn't a good idea is the complexity it adds to the system. When the publishing transaction is _assumed_ to exist, it is known on application start, provided as configuration. However, doing it part of the init means that there needs to be code that (a) realize the event and (b) monitor the outcome of the event. That's extra complexity for the Hydra node which we may want to prevent.

Author: Good points. You are right in that the DoS stuff is about spendability of the outputs. In option 1 we could make it spendable by head instance itself and only when terminating it, i.e. by ensuring the ST is also burned. So it can be made DoS-safe and non-littering. Which is not possible in option 2.

* We will publish scripts on release of the `hydra-node`, or more specifically of the `hydra-plutus` package.

* The scripts are stored at outputs addressed to some **unspendable** `v_publish` validator.

>   + This is to avoid DoS risk and unnecessariy centralization with option 2

Reviewer: I was thinking that, we could imagine a `v_genesis_delegate` which can only be spent with a quorum of 5 signatures from the genesis delegates :sweat_smile: ... As a "standard", this would allow to create scripts that can be later garbage-collected by the genesis delegates to reduce littering of the chain. They can already DoS the entire chain if they wanted to, so letting them also get rid of the UTxO if needs be sounds fair enough :sweat_smile: 

I guess it's okay anyway to have a `const False` script because we'll be talking about few scripts a year :shrug: ... still, somehow it bothers me to "leave my trash around".

## Consequences

* We need a process and/or tool to publish `hydra-plutus` scripts and need to pay the deposits.

>   + Any other party could to the same, this does not lead to centralization.

Reviewer: I've got a proof of concept in mind, little web app connecting to a CIP-0030's wallet and using the Github API to fetch release artifacts :ok_hand:

> * The `hydra-node` would be need to know the `TxIn`s of the "right" published scripts. In the simplest case we would just make this configurable.

Reviewer: I'd park that as an "implementation details" (the `TxIn` bit). As we discussed, we need to provide the Hydra node with the output reference corresponding to each scripts needed as reference. This can be inferred from a `TxIn` + state query (as we currently do), or could be specified as a configuration file / command-line options fully.

Author: Are you saying that it's okay to leave it like this and how the `TxIn` gets inferred is an implementation detail? i.e. not enumerating all but doing what we do right now is an optimization on convenience == implementation detail not covered by the adr

Reviewer: I mean that mentioning `TxIn` is an implementation detail. The ADR should simply say that a reference to the published scripts must be communicated to the Hydra node. An address, a TxId, a TxIn ... doesn't matter much.

Reviewer: Perhaps I am just neat picking.

Author: Ah well. Ultimately the reference needs to be available as a `TxIn` eventually, so I think that formulation is the common denominator all specific implementations need to fulfill. :bike: :house:

> * If we parameterize the `v_publish` validator by a "type + version" indicator, this yield a unique address per Head script version.

Reviewer: We'd likely want to also include some kind of project identifier in there. Because we could imagine this approach to be "normalized" and reused. Alternatively, we could also use a static address and use inline-datum as a way to attach metadata. Then it's possible to discover all scripts published in that way and tell them apart from the metadata.

Author: Yeah maybe better to store the tags in datums. In this specific case the examples should highlight that any tag should be okay and we use a script name + version in `HydraHeadV1`, `HydraInitialV1` etc.

* The general trade-off is: instead of paying ADA fees for scripts adding to the transaction size in _each_ transaction, ADA deposits will need to be put down to have scripts be part of the UTxO set in the ledger _once_.

## Decision

* Publish outputs holding Hydra scripts out-of-band (option 2), because
  + All scripts would not fit into the `init` transaction directly, we would need to post multiple.
  + Costs (deposits) would need to be payed for each head instance.

* The scripts are stored at outputs addressed to some **unspendable** `v_publish` validator.
  + This is to avoid DoS risk and unnecessariy centralization
  + We have considered "garbage collection" by allowing spending these outputs into re-publishing new versions of the script.
    - This would make things even more complicated and we decided to not bother about "littering the chain" right now.

* We will publish scripts on release of the `hydra-node`, or more specifically of the `hydra-plutus` package.

## Consequences

* We need a process and/or tool to publish `hydra-plutus` scripts and need to pay the deposits.

>   + Any other party could to the same, this does not lead to centralization.

Reviewer 2: ```suggestion
  + Any other party could do the same, this does not lead to centralization.
```
